packages/model-fragments/lib/fragments/ext.js

```diff
@@ -152,83 +100,141 @@ Model.reopen({
   */
   changedAttributes: function() {
     var diffData = this._super();
+    var internalModel = internalModelFor(this);
 
-    Ember.keys(this._fragments).forEach(function(name) {
+    Ember.keys(internalModel._fragments).forEach(function(name) {
       // An actual diff of the fragment or fragment array is outside the scope
       // of this method, so just indicate that there is a change instead
-      if (name in this._attributes) {
+      if (name in internalModel._attributes) {
         diffData[name] = true;
       }
     }, this);
 
     return diffData;
   },
+});
 
-  /**
-    If the model `isDirty` this function will discard any unsaved
-    changes, recursively doing the same for all fragment properties.
+// Replace a method on an object with a new one that calls the original and then
+// invokes a function with the result
+function decorateMethod(obj, name, fn) {
```

Comment:
@igorT the diff is a little difficult to read, but [these frankenstein method decorators](https://github.com/lytics/ember-data.model-fragments/blob/internal-model-compat/packages/model-fragments/lib/fragments/ext.js#L117-L195) are one of the terrible things I mentioned.

I tried pretty hard to stay away from touching `InternalModel`'s prototype, but was unable to accomplish what these three methods do without doing so:

1. `createSnapshot` - since fragment attributes are not relationships, they are accessed through `Snapshot#attr`, which means that by the time the user sees a record snapshot, its fragment attributes must already be snapshotted. The alternative was doing something similar to `Snapshot`'s prototype...
2. `rollbackAttributes` - This could potentially have stayed in the `Model.reopen()`, but then it runs the risk of getting out of sync if `InternalModel#rollbackAttributes` is ever [called directly](https://github.com/emberjs/data/blob/23a5b13a9bccd3f26515f4992dd59960d6649df0/packages/ember-data/lib/system/model/states.js#L276).
3. `adapterDidCommit` - It doesn't look like there's any other hook where updating fragment states can happen. This one is particularly nasty because of the need to use the original method's `data` argument.